homeassistant/components/zwave_js/services.py

```diff
@@ -0,0 +1,216 @@
+"""Methods and classes related to executing Z-Wave commands and publishing these to hass."""
+import logging
+from typing import Dict, List, Union, cast
+
+import voluptuous as vol
+from zwave_js_server.const import CommandClass
+from zwave_js_server.model.node import Node as ZwaveNode, get_value_id
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_DEVICE_ID, ATTR_ENTITY_ID
+from homeassistant.core import HomeAssistant, ServiceCall, callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.device_registry import DeviceRegistry
+from homeassistant.helpers.entity_registry import EntityRegistry
+
+from . import const
+from .entity import get_home_and_node_id_from_device_id
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def convert_bitmask_to_int(value: str) -> int:
+    """Convert a bitmask (hex) to an integer."""
+    return int(value, 16)
+
+
+def parameter_name_does_not_need_bitmask(
+    val: Dict[str, Union[int, str]]
+) -> Dict[str, Union[int, str]]:
+    """Validate that if a parameter name is provided, bitmask is not as well."""
+    if isinstance(val[const.ATTR_CONFIG_PARAMETER], str) and (
+        val.get(const.ATTR_CONFIG_PARAMETER_BITMASK)
+    ):
+        raise vol.Invalid(
+            "Don't include a bitmask when a parameter name is specified",
+            path=[const.ATTR_CONFIG_PARAMETER, const.ATTR_CONFIG_PARAMETER_BITMASK],
+        )
+    return val
+
+
+BITMASK_SCHEMA = vol.All(
+    cv.string, vol.Lower, vol.Match(r"^(0x)?[0-9a-f]+$"), convert_bitmask_to_int
+)
+
+
+class ZWaveServices:
+    """Class that holds our services (Zwave Commands) that should be published to hass."""
+
+    def __init__(
+        self, hass: HomeAssistant, dev_reg: DeviceRegistry, ent_reg: EntityRegistry
+    ):
+        """Initialize with hass, device registry, and entity registry objects."""
+        self._hass = hass
+        self._dev_reg = dev_reg
+        self._ent_reg = ent_reg
+
+    @callback
+    def async_register(self) -> None:
+        """Register all our services."""
+        self._hass.services.async_register(
+            const.DOMAIN,
+            const.SERVICE_SET_CONFIG_PARAMETER,
+            self.async_set_config_value,
+            schema=vol.All(
+                {
+                    vol.Exclusive(ATTR_DEVICE_ID, "id"): cv.string,
+                    vol.Exclusive(ATTR_ENTITY_ID, "id"): cv.entity_ids,
+                    vol.Optional(const.ATTR_ENDPOINT): vol.Coerce(int),
```

Comment:
Is the endpoint needed ? Isn't it true that all configuration parameters are on endpoint 0 (the node itself) ?

Reply:
is that the case? if so we can get rid of this, I looked around at fixtures and could only find these values on endpoint 0 but I wasn't sure if it was a hard rule

Reply:
Configuration settings are node specific so I think it's safe to drop.